staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go

```diff
@@ -552,16 +652,45 @@ func (alloc *allocator) allocateOne(r deviceIndices) (bool, error) {
 		return alloc.allocateOne(deviceIndices{claimIndex: r.claimIndex + 1})
 	}
 
+	requestKey := requestIndices{claimIndex: r.claimIndex, requestIndex: r.requestIndex, subRequest: r.subRequest, subRequestIndex: r.subRequestIndex}
+	requestData := alloc.requestData[requestKey]
+	// If the request has subRequests, attempt them in order until one can be satisfied
+	// or all have been attempted.
+	if requestData.hasSubRequests {
+		for i := 0; ; i++ {
+			_, ok := alloc.requestData[requestIndices{claimIndex: r.claimIndex, requestIndex: r.requestIndex, subRequest: true, subRequestIndex: i}]
+			if !ok {
+				return false, nil
+			}
+			success, err := alloc.allocateOne(deviceIndices{claimIndex: r.claimIndex, requestIndex: r.requestIndex, subRequest: true, subRequestIndex: i})
```

Comment:
```suggestion
			// r.subRequest must be false, otherwise requestData.hasSubRequests wouldn't be true.
			// Setting `subRequest: true` for the next recursion will
			// try to allocate that specific subrequest instead of entering this for loop.
			success, err := alloc.allocateOne(deviceIndices{claimIndex: r.claimIndex, requestIndex: r.requestIndex, subRequest: true, subRequestIndex: i})
```